File: 1_algorithmic_toolbox/week4/cpp/3_majority_element.cpp
#include <algorithm>
#include <iostream>
#include <vector>

bool majority_element(std::vector<int> &sequence)
{
  int n_maj { -1 }, curr { -1 }, n_curr { -1 };

  std::sort(sequence.begin(), sequence.end());
  for (const auto &e : sequence)
    if (e == curr)
      ++n_curr;
    else {
      n_maj  = std::max(n_maj, n_curr);
      curr   = e;
      n_curr = 1;
    }

  return std::max(n_maj, n_curr) > sequence.size() / 2;
}

int main()
{
  int n {};
  std::cin >> n;
  std::vector<int> sequence(n);
  for (auto &e : sequence) std::cin >> e;

  std::cout << majority_element(sequence) << '\n';
}

File: 1_algorithmic_toolbox/week4/rust/3_majority_element.rs
fn main() {
    let mut buffer = String::new();
    let sequence = line_reader(&mut buffer, 2).skip(1).collect();
    println!("{}", majority_element(sequence) as i32);
}

fn line_reader<'a, T>(buffer: &'a mut String, n_lines: usize) -> Box<dyn Iterator<Item = T> + 'a>
where
    T: std::str::FromStr,
    <T as std::str::FromStr>::Err: std::fmt::Debug,
{
    buffer.clear();
    for _ in 0..n_lines {
        std::io::stdin().read_line(buffer).unwrap();
    }

    Box::new(buffer.split_whitespace().map(|i| i.parse().unwrap()))
}

fn majority_element(mut sequence: Vec<i32>) -> bool {
    let n = sequence.len() as i32;
    let (mut n_maj, mut curr, mut n_curr) = (-1, -1, -1);

    sequence.sort();
    sequence.into_iter().for_each(|e| {
        if e != curr {
            n_maj = std::cmp::max(n_maj, n_curr);
            (curr, n_curr) = (e, 0);
        }
        n_curr += 1;
    });

    std::cmp::max(n_maj, n_curr) > n / 2
}
